### ISSUE TITLE
BufferTooSmall in gcp_cloud_storage sink

### ISSUE THREAD

**Issue by gschier** [vectordotdev/vector#13102]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

We're periodically seeing a `BufferTooSmall` error in our gcp_storage sink. Note, this also caused an exit 0 but it seems like it should exit with a non-zero code in this case?

```
Jun 10 17:08:17 production-logger vector[183068]: thread 'vector-worker' panicked at 'writer hit unrecoverable error during write: failed to encode record: BufferTooSmall', /project/lib/vector-buffers/src/topology/channel/sender.rs:63:21
Jun 10 17:08:17 production-logger vector[183068]: note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Jun 10 17:08:17 production-logger vector[183068]: thread 'vector-worker' panicked at 'join error: JoinError::Panic(Id(1518), ...)', src/topology/builder.rs:212:30
Jun 10 17:08:17 production-logger vector[183068]: 2022-06-10T17:08:17.247685Z ERROR source{component_kind="source" component_id=in_vector component_type=vector component_name=in_vector}: vector::topology: An error occurred that Vector couldn't handle.
Jun 10 17:08:17 production-logger vector[183068]: 2022-06-10T17:08:17.247959Z  INFO vector: Vector has stopped.
Jun 10 17:08:17 production-logger vector[183068]: 2022-06-10T17:08:17.248542Z  INFO vector::topology::running: Shutting down... Waiting on running components. remaining_components="out_datadog, out_file, out_gcp_storage, in_vector" time_remaining="59 seconds left"
Jun 10 17:08:22 production-logger vector[183068]: 2022-06-10T17:08:22.249686Z  INFO vector::topology::running: Shutting down... Waiting on running components. remaining_components="out_gcp_storage" time_remaining="54 seconds left"
Jun 10 17:08:27 production-logger vector[183068]: 2022-06-10T17:08:27.249528Z  INFO vector::topology::running: Shutting down... Waiting on running components. remaining_components="out_gcp_storage" time_remaining="49 seconds left"
Jun 10 17:08:32 production-logger vector[183068]: 2022-06-10T17:08:32.248993Z  INFO vector::topology::running: Shutting down... Waiting on running components. remaining_components="out_gcp_storage" time_remaining="44 seconds left"
Jun 10 17:08:37 production-logger vector[183068]: 2022-06-10T17:08:37.248905Z  INFO vector::topology::running: Shutting down... Waiting on running components. remaining_components="out_gcp_storage" time_remaining="39 seconds left"
Jun 10 17:08:42 production-logger vector[183068]: 2022-06-10T17:08:42.248771Z  INFO vector::topology::running: Shutting down... Waiting on running components. remaining_components="out_gcp_storage" time_remaining="34 seconds left"
Jun 10 17:08:47 production-logger vector[183068]: 2022-06-10T17:08:47.249258Z  INFO vector::topology::running: Shutting down... Waiting on running components. remaining_components="out_gcp_storage" time_remaining="29 seconds left"
Jun 10 17:08:52 production-logger vector[183068]: 2022-06-10T17:08:52.248895Z  INFO vector::topology::running: Shutting down... Waiting on running components. remaining_components="out_gcp_storage" time_remaining="24 seconds left"
Jun 10 17:08:57 production-logger vector[183068]: 2022-06-10T17:08:57.249167Z  INFO vector::topology::running: Shutting down... Waiting on running components. remaining_components="out_gcp_storage" time_remaining="19 seconds left"
Jun 10 17:09:02 production-logger vector[183068]: 2022-06-10T17:09:02.249650Z  INFO vector::topology::running: Shutting down... Waiting on running components. remaining_components="out_gcp_storage" time_remaining="14 seconds left"
Jun 10 17:09:07 production-logger vector[183068]: 2022-06-10T17:09:07.249386Z  INFO vector::topology::running: Shutting down... Waiting on running components. remaining_components="out_gcp_storage" time_remaining="9 seconds left"
Jun 10 17:09:12 production-logger vector[183068]: 2022-06-10T17:09:12.248860Z  INFO vector::topology::running: Shutting down... Waiting on running components. remaining_components="out_gcp_storage" time_remaining="4 seconds left"
Jun 10 17:09:17 production-logger vector[183068]: 2022-06-10T17:09:17.249546Z ERROR vector::topology::running: Failed to gracefully shut down in time. Killing components. components="out_gcp_storage"
Jun 10 17:09:18 production-logger systemd[1]: railwaylogaggregator.service: Succeeded.
Jun 10 17:09:18 production-logger systemd[1]: railwaylogaggregator.service: Consumed 9h 25min 8.969s CPU time.
```

### Configuration

```text
data_dir = "${LOGGER_VECTOR_PATH}"

[api]
enabled = true
playground = false

[sources.in_vector]
type = "vector"
version = "2"
address = "0.0.0.0:5143"

[sinks.out_file]
type = "file"
inputs = ["in_vector"]
compression = "none"
encoding = "ndjson"
path = "${LOGGER_VECTOR_PATH}/out_file/logs/%F_%H/{{ appname }}.log"

[sinks.out_gcp_storage]
type = "gcp_cloud_storage"
inputs = ["in_vector"]
bucket = "${LOGGER_BUCKET_NAME:-railway-develop-logs}"
key_prefix = "logs/{{ appname }}/"
compression = "none"
buffer.type = "disk"
buffer.max_size = 1_500_000_000_000
batch.timeout_secs = 600
batch.max_bytes = 10_000_000
filename_append_uuid = false
encoding = "ndjson"
# Note: Time in `filename_time_format` uses time-of-last-log unlike `key_prefix` which uses time-of-first-log
filename_time_format = "%F/%T"

# Internal metrics for our reporting to DataDog

[sources.in_internal]
type = "internal_metrics"
namespace = "vector"
scrape_interval_secs = 60

[transforms.transform_datadog]
type = "filter"
inputs = [ "in_internal" ]
condition = '.name == "utilization" || .name == "component_sent_event_bytes_total" || .name == "component_received_event_bytes_total"'

[sinks.out_datadog]
type = "datadog_metrics"
inputs = ["transform_datadog"]
api_key = "${DD_API_KEY}"
```

### Version

vector 0.22.0 (x86_64-unknown-linux-gnu 5e937e3 2022-06-01)

**Comment by spencergilbert** [comment]
cc @tobz

**Comment by jszwedko** [comment]
Thanks for this report @gschier ! We have some ideas for remedying it already. As a work around, if this is blocking you, you can opt back into the old buffer implementation via `buffer.type = "v1"`. Agreed that Vector should also be exiting non-zero here.

**Comment by tobz** [comment]
Just to note, however: doing so on an existing host will create a _new_ buffer. The current buffer won't be converted back to the old format.

**Comment by tobz** [comment]
FYI, we're looking to address this via https://github.com/vectordotdev/vector/pull/13121 which will increase the maximum record size by 16x, from 8MB to 128MB.

This should hopefully be large enough for the batch sizes you're using, but if not, you would still have one escape hatch left which would be tweaking your batch size on the `vector` sink side to a lower value. That option is actually open and available to you now, even, but the above PR should hopefully be large enough to let you avoid having to do so, which would conceivably affect performance in your set-up given the scale of logging.

**Comment by jszwedko** [comment]
Should be closed by https://github.com/vectordotdev/vector/pull/13121

**Comment by jszwedko** [comment]
This'll be included in 0.22.2.

**Comment by gschier** [comment]
Amazing news! Thanks for the update 👍🏻 🙂
